### PR TITLE
feat: Allow mypy to check untyped defs

### PR DESCRIPTION
**Pull request by tachikoma-li**
To set `check_untyped_defs` to `True`, we need to make sure not to change the type of variable within the function body.
This PR is to fix that issue. 

The main change is in `_qctrl_expanded_export_content` where the type of return value `control_info` depends on the `coordinates`. It is a `List` for CSV and `Dict` for JSON. However, we can always make `control_info` as a `Dict` and change the format is the `_export_to_qctrl_expanded_format` (which calls `_qctrl_expanded_export_content` to get `control_info`) which only save the data to file and therefore does not have a return type but only side effect.

some other minor tweaks:
- make `filename` mandatory
- minor tweak  in `export_to_file` to fix the type issue and remove the error-rasing part
- minor tweak in `new_ramsey_sequence` to fix the type issue

### CODE REVIEW

File: qctrlopencontrols/driven_controls/driven_control.py [hunk 1]
```diff
@@ -425,21 +390,28 @@ def _export_to_qctrl_expanded_format(
...
+                        {name: control_info[name][index] for name in field_names}
+                    )
         else:
             with open(filename, "wt") as handle:
                 json.dump(control_info, handle, sort_keys=True, indent=4)
 
     def export_to_file(
         self,
-        filename=None,
+        filename,
```
Comment (charmasaur) [inline qctrlopencontrols/driven_controls/driven_control.py:414]: Update docstring

File: qctrlopencontrols/driven_controls/driven_control.py [hunk 2]
```diff
@@ -327,119 +330,88 @@ def duration(self) -> float:
...
         Prepare the content to be saved in Q-CTRL expanded format.
 
         Parameters
         ----------
-        file_type : str, optional
-            One of 'CSV' or 'JSON'; defaults to 'CSV'.
         coordinates : str, optional
             Indicates the co-ordinate system requested. Must be one of
-            'cylindrical', 'cartesian' or 'polar'; defaults to 'cylindrical'
+            'cylindrical', 'cartesian' or 'polar'. Defaults to 'cylindrical'.
```
Comment (tachikoma-li) [inline qctrlopencontrols/driven_controls/driven_control.py:341]: BTW, I don't think we have the `polar` option, shall we remove it?

File: qctrlopencontrols/driven_controls/driven_control.py [hunk 3]
```diff
@@ -327,91 +331,52 @@ def duration(self) -> float:
...
+        else:
+            control_info["rabi_rates"] = list(self.rabi_rates / self.maximum_rabi_rate)
+            control_info["azimuthal_angles"] = list(self.azimuthal_angles)
 
         return control_info
 
     def _export_to_qctrl_expanded_format(
         self,
-        filename=None,
+        filename,
```
[the remark this answers is not in the document]
Comment (tachikoma-li) [reply qctrlopencontrols/driven_controls/driven_control.py:373]: done

File: qctrlopencontrols/driven_controls/driven_control.py [hunk 4]
```diff
@@ -425,21 +390,28 @@ def _export_to_qctrl_expanded_format(
...
-            with open(filename, "wt") as handle:
-
-                control_info = "\n".join(control_info)
-                handle.write(control_info)
+            _ = control_info.pop("name")
+            control_info["maximum_rabi_rate"] = [
+                self.maximum_rabi_rate
+            ] * self.number_of_segments
+            field_names = sorted(control_info.keys())
+            with open(filename, "w", newline="") as file:
```
[the remark this answers is not in the document]
Comment (tachikoma-li) [reply qctrlopencontrols/driven_controls/driven_control.py:401]: Good point. Added.

File: qctrlopencontrols/driven_controls/driven_control.py (hunk 1 above)
Comment (tachikoma-li) [reply qctrlopencontrols/driven_controls/driven_control.py:414]: Sorry, did you mean the doc for `filename`?
Seems it was not optional in the doc.

File: qctrlopencontrols/driven_controls/driven_control.py (hunk 2 above)
Comment (charmasaur) [reply qctrlopencontrols/driven_controls/driven_control.py:341]: Yup, I'd guess that's what we used to call cylindrical.

File: qctrlopencontrols/driven_controls/driven_control.py (hunk 1 above)
Comment (charmasaur) [reply qctrlopencontrols/driven_controls/driven_control.py:414]: D'oh, sorry. I just assumed it was.

**Review by charmasaur — COMMENTED** [review]
LGTM, although feel free to remove "polar" too.

**Review by charmasaur — APPROVED** [review]

**Pull request merged**